sqlmesh/core/loader.py

```diff
@@ -208,7 +210,17 @@ def _load_external_models(self) -> UniqueKeyDict[str, Model]:
                             project=config.project,
                             default_catalog=self._context.default_catalog,
                         )
+                        external_models.append(model)
+
+                    # external models with no explicit gateway defined form the base set
+                    for model in (e for e in external_models if e.gateway is None):
```

Comment:
I wonder whether this can be simplified to one pass through:
```
for model in external_models:
    if model.fqn not in models or model.gateway == gateway:
        models[model.fqn] = model
```
Does this achieve the same thing?

Reply:
Almost. That achieves:

```ValueError: Duplicate key found in UniqueKeyDict<models>. Call dict.update(...) if this is intentional.```

However, if I change it to `models.update({...})` then it still breaks a couple of tests because it adds external models from other gateways (not just `None` and the current gateway).

If it change it to:
```
                    for model in external_models:
                        if model.gateway is None or model.gateway == gateway:
                            models.update({model.fqn: model})
```

Then there is the potential to get different results depending on which order the models are loaded. For example, a model with `gateway=None` could override a model with the same name and `gateway=<current gateway>` if it's loaded second.

The two-pass method avoids this by populating the base set (models with `gateway=None`) and then overriding it just with models from the current gateway.

I get that it's less efficient to iterate the list twice. If I tried to sort the list (to put all the `gateway=None` entries at the top) and iterate it once, I guess that's still iterating it twice because the sort will need to iterate it

Reply:
Oh, if you're getting:
> ValueError: Duplicate key found in UniqueKeyDict<models>. Call dict.update(...) if this is intentional.

Doesn't this mean that your previous implementation doesn't work either, since no override actually takes place?

Reply:
```
for model in external_models:
    if model.fqn not in models or (model.gateway and model.gateway == gateway):
        models.update({model.fqn: model})
```
Shouldn't this work then?

Reply:
Anyway, this is a minor thing.

Reply:
> Doesn't this mean that your previous implementation doesn't work either, since no override actually takes place?

The overrides take place on the second pass, where `update` is called instead of using the `foo[bar] = "baz"` syntax. This ensures that duplicate models without a gateway defined will still cause an error (current behaviour)

> Shouldn't this work then?

It doesnt, because the first check `if model.fqn not in models` can return `True` for models specific to an unrelated gateway. 

Let's say I had a model `@{gateway}_foo` and had two versions of it defined in the external models file:

```
- name: dev_foo
  gateway: dev
- name: prod_foo
  gateway: prod
```

Using your code, loading this file using `--gateway dev` will put `prod_foo` in the list of external models, because `model.fqn not in models` returns `True`

Does that make sense? You can try it as well, some tests that check for this specifically will fail